### ISSUE TITLE
To use ACTIONS_RUNTIME_TOKEN for writing into Package Registry

### ISSUE THREAD

**Issue by mstup** [go-gitea/gitea#29398]
### Feature Description

Because ACTIONS_RUNTIME_TOKEN could be used to pull the code from Repository, it looks obvious to use this token later to write build result into Package Registry.

**Comment by lunny** [comment]
Pulling the code from the Repository is a read permission of Code, it's not a write permission. I don't think the logic is reasonable.

**Comment by mstup** [comment]
I think that end user should be able to make this decision. Because hardcoding or mocking some dummy user looks even less reasonable.

**Comment by mstup** [comment]
This is screenshot from GitHub page and looks like they agreed with that.

![image](https://github.com/go-gitea/gitea/assets/16793676/d8c308b6-d502-4515-a6bd-095fc0001661)

**Comment by mstup** [comment]
GitLab as well propose as option to use [${CI_JOB_TOKEN}](https://docs.gitlab.com/ee/user/packages/maven_repository/#basic-http-authentication) to access to Maven Package Registry.

**Comment by mstup** [comment]
@ChristopherHX As I see it, **ACTIONS_RUNTIME_TOKEN** regenerates each time on job starts execution. It looks much more secure instead of any hard coded value. No sense to store and remember this at all.
BTW: Values of **ACTIONS_RUNTIME_TOKEN** and **GITHUB_TOKEN** are equals now (checked it on Gitea v1.21.7)

**Comment by mstup** [comment]
Looks like this [issue 23642](https://github.com/go-gitea/gitea/issues/23642) is about the same problem.

**Comment by Frankkkkk** [comment]
I agree, this would be really nice and remove the need for PATs. The way Gitlab does it is quite neat IMHO :)

**Comment by vizh** [comment]
It seems that for each project (many many projects), we now need to create ubique dummy user and two tokens for it:

1. will be hardcoded in CI and used in the build process to upload packages.
2. will be hardcoded on the test and production servers to download packages.

This results in the creation of a large number of extra users, which will interfere with regular users using the interfaces, as they cannot be hidden.

Consequently, the system will have hundreds of dummy users. Creating just a few and grant access to multiple repositories is not correct from a security standpoint. Additionally, we need to remember their passwords to log in and generate tokens.

Perhaps a good solution would be an ability to create tokens for projects rather than users. This way, there would be no extra users in the system.

**Comment by thequux** [comment]
I don't consider using a PAT as a secret to be a complete solution to this problem for all of the reasons above (profusion of users, risk of long-lived tokens leaking, etc). Further, the problem is more general than just writing to the package registry. My own use case is that I am writing an action that builds a preview of a static site and uploads it to a public preview with an unguessable URL to make it easy for stakeholders to provide comments on the update while mitigating the risk of information leaks. Given that the random URL is designed to be unguessable, my action should push a comment to the PR with the random URL.

I propose to solve this problem in an extensible way by making the ACTIONS_RUNTIME_TOKEN have configurable permissions per repository. Ideally, it would be possible to restrict the token to only have access to the repo that's running the PR, or even the PR itself.

Having reviewed the existing authorization infrastructure in gitea, the basic version of this would be fairly easy to add via a mapping from repository to scopes granted to actions. The more complete solution, where scopes can be limited to specific resources, will require more elaborate changes to the concept of an access token scope, but I can see a fairly straightforward way to implement this without significant overhead.

If this sounds like a good idea to the Gitea team, I'll put the effort in to make a PR. I can handle the API design and backend changes myself, but I'll need help from a UX designer to build a user-facing interface.

**Comment by Frankkkkk** [comment]
@thequux this issue could be of interest to you:  [Proposal] Support configuring permissions of automatic tokens of Actions jobs #24635

**Comment by ChristopherHX** [comment]
Since I noticed recently ACTIONS_RUNTIME_TOKEN and gitea.token are now interchangeble again, this token could indeed get package registry access.

---

IMO Problems that needs to be addressed by package registry
- currently there is no permission concept per package, so the permission middleware runs without knowledge of the package name is not so trivial
  - allow gitea.token to create new packages if it has `package: write` scope (pr from fork get max read access, if not using privileged workflows)
    - automatically assign actions permission
  - allow repositories that has assigned actions write acces in package settings

@thequux 

> I propose to solve this problem in an extensible way by making the ACTIONS_RUNTIME_TOKEN have configurable permissions per repository. Ideally, it would be possible to restrict the token to only have access to the repo that's running the PR, or even the PR itself.

This sound like a different topic, please take this as reference <https://github.com/go-gitea/gitea/pull/32562>

Since this is more similar to your proposal, for access to repositories. Difference are you want write acces level instead of read and do want to use it from a public repository.
